File: test/js/app-control-gallery.js
// Interactive draw/behavior test for the PDG JavaScript application controls.
// Run with: ./pdg test/js/app-control-gallery.js

if (typeof pdg === 'undefined') global.pdg = require('pdg');
// Log the first three native drawText argument sets for every gallery button.
// This gallery is a diagnostic executable, so keep the evidence in its output.
global.PDG_CONTROL_DRAW_DIAGNOSTICS = { maxDrawsPerButton: 3 };
const framework = require('../../src/js/mvc-app');
const { Application } = framework.Application;
const { Controller } = framework.Controller;
const { View } = framework.View;
const { Button } = framework.Button;
const { Checkbox } = framework.Checkbox;
const { RadioButton } = framework.RadioButton;
const { Scrollbar, ScrollbarOrientation } = framework.Scrollbar;
const { Dialog, DialogFlags } = framework.Dialog;
const {
    ControlAttributes,
    ControlState,
    ControlType
} = framework.ControlAttributes;

const ids = {
    defaultButton: 100,
    disabledButton: 101,
    themedButton: 102,
    imageButton: 103,
    defaultDialog: 104,
    themedDialog: 105
};

class GalleryCanvas extends View {
    drawSelf(port) {
        port.drawRect(this.getViewArea(), new pdg.Attributes()
            .fillColor(new pdg.Color(0.94, 0.95, 0.97, 1)));
        port.drawText('PDG JavaScript App Framework Control Gallery', new pdg.Point(40, 48),
            new pdg.Attributes().textSize(25).textStyle(pdg.textStyle_Bold)
                .fillColor(new pdg.Color(0.12, 0.15, 0.22, 1)));
        port.drawText('Hover, press, click, toggle, and open both dialogs.', new pdg.Point(40, 76),
            new pdg.Attributes().textSize(14).fillColor(new pdg.Color(0.28, 0.31, 0.36, 1)));
        port.drawRect(new pdg.Rect(30, 92, 450, 500), new pdg.Attributes()
            .fillColor(new pdg.Color(1, 1, 1, 1)).lineColor(new pdg.Color(0.73, 0.75, 0.79, 1))
            .roundedCorners(10));
        port.drawRect(new pdg.Rect(510, 92, 930, 500), new pdg.Attributes()
            .fillColor(new pdg.Color(0.98, 0.97, 1, 1)).lineColor(new pdg.Color(0.49, 0.43, 0.71, 1))
            .lineThickness(2).roundedCorners(10));
        port.drawText('Built-in defaults', new pdg.Point(50, 116),
            new pdg.Attributes().textSize(17).textStyle(pdg.textStyle_Bold));
        port.drawText('Per-control overrides', new pdg.Point(530, 116),
            new pdg.Attributes().textSize(17).textStyle(pdg.textStyle_Bold)
                .fillColor(new pdg.Color(0.29, 0.22, 0.57, 1)));
        port.drawRect(new pdg.Rect(30, 525, 930, 590), new pdg.Attributes()
            .fillColor(new pdg.Color(0.13, 0.15, 0.21, 1)).roundedCorners(8));
        port.drawText(`Behavior: ${this.controller.status}`, new pdg.Point(50, 557),
            new pdg.Attributes().textSize(16).fillColor(new pdg.Color(1, 1, 1, 1)));
        port.drawText('Overrides: state colors, image, draw routine, click routine, and dialog theme.',
            new pdg.Point(50, 580), new pdg.Attributes().textSize(12)
                .fillColor(new pdg.Color(0.75, 0.8, 0.88, 1)));
    }
}

class DialogLabel extends View {
    constructor(controller, area, text) {
        super(controller, area);
        this.text = text;
    }

    drawSelf(port) {
        port.drawText(this.text, new pdg.Point(this.viewArea.left + 12, this.viewArea.top + 28),
            new pdg.Attributes().textSize(17).fillColor(new pdg.Color(0, 0, 0, 1)));
    }
}

class PreviewDialog extends Dialog {
    constructor(parent, themed) {
        super(parent, 360, 150, DialogFlags.dialog_Standard, 1);
        new DialogLabel(this, this.getDialogRect(), themed
            ? 'Custom draw routine for dialog background'
            : 'Default PDG dialog background');
        const area = this.getDialogRect();
        const close = new Button(this,
            new pdg.Rect(area.right - 112, area.bottom - 44, area.right - 22, area.bottom - 14), 1);
        close.setText('Close');
        close.setID(1);
    }
}

class GalleryController extends Controller {
    setupViews() {
        this.status = 'Click any enabled control';
        this.useThemedDialog = false;
        this.canvas = new GalleryCanvas(this, this.port.getDrawingArea());

        this.addButton(new pdg.Rect(55, 125, 225, 165), ids.defaultButton, 'Default button');
        const disabled = this.addButton(new pdg.Rect(55, 180, 225, 220), ids.disabledButton, 'Disabled');
        disabled.setEnabled(false);

        const themedButton = new ControlAttributes()
            .stateDrawRoutine(ControlState.Normal, GalleryController.drawAccentButton)
            .stateDrawRoutine(ControlState.Hovered, GalleryController.drawAccentButton)
            .stateDrawRoutine(ControlState.Pressed, GalleryController.drawPressedAccentButton)
            .stateForeground(ControlState.Normal, new pdg.Color(1, 1, 1, 1))
            .stateForeground(ControlState.Hovered, new pdg.Color(1, 1, 1, 1))
            .stateForeground(ControlState.Pressed, new pdg.Color(1, 1, 1, 1))
            .clickRoutine(() => { this.status = 'Custom button click routine ran'; });
        this.addButton(new pdg.Rect(535, 125, 705, 165), ids.themedButton, 'Draw routine', themedButton);

        let exampleImage = null;
        try { exampleImage = this.app.getResourceManager().getImage('yinyang.png'); } catch (_) {}
        const imageButton = new ControlAttributes()
            .stateImage(ControlState.Normal, exampleImage)
            .stateForeground(ControlState.Normal, new pdg.Color(1, 1, 1, 1));
        this.addButton(new pdg.Rect(730, 125, 900, 165), ids.imageButton, 'Image state', imageButton);

        this.addCheckbox(new pdg.Rect(55, 245, 360, 277), 'Default checkbox');
        const disabledCheck = this.addCheckbox(new pdg.Rect(55, 277, 360, 309), 'Disabled checkbox');
        disabledCheck.setEnabled(false);
        this.addCheckbox(new pdg.Rect(535, 245, 850, 277), 'Override text colors',
            new ControlAttributes()
                .stateForeground(ControlState.Normal, new pdg.Color(0.18, 0.27, 0.55, 1))
                .stateForeground(ControlState.Selected, new pdg.Color(0.64, 0.19, 0.36, 1))
                .clickRoutine(() => { this.status = 'Custom checkbox toggled'; }));

        this.addRadio(new pdg.Rect(55, 315, 370, 345));
        const disabledRadio = this.addRadio(new pdg.Rect(55, 355, 370, 385));
        disabledRadio.setEnabled(false);
        this.addRadio(new pdg.Rect(535, 315, 850, 345), new ControlAttributes()
            .stateForeground(ControlState.Normal, new pdg.Color(0.14, 0.37, 0.28, 1))
            .stateForeground(ControlState.Selected, new pdg.Color(0.75, 0.29, 0.12, 1))
            .clickRoutine(() => { this.status = 'Custom radio selection changed'; }));

        new Scrollbar(this, new pdg.Rect(55, 402, 370, 424),
            ScrollbarOrientation.HORIZONTAL, 35, 10, 110);
        const themedScrollbar = new Scrollbar(this, new pdg.Rect(535, 402, 850, 424),
            ScrollbarOrientation.HORIZONTAL, 65, 10, 110);
        themedScrollbar.setAttributes(new ControlAttributes()
            .stateAttributes(ControlState.Normal, new pdg.Attributes().fillColor(new pdg.Color(0.88, 0.85, 0.96, 1)))
            .stateAttributes(ControlState.Decrement, new pdg.Attributes().fillColor(new pdg.Color(0.53, 0.47, 0.81, 1)).roundedCorners(4))
            .stateAttributes(ControlState.Increment, new pdg.Attributes().fillColor(new pdg.Color(0.53, 0.47, 0.81, 1)).roundedCorners(4))
            .stateAttributes(ControlState.Thumb, new pdg.Attributes().fillColor(new pdg.Color(0.64, 0.19, 0.36, 1)).roundedCorners(6)));

        this.addButton(new pdg.Rect(55, 447, 255, 487), ids.defaultDialog, 'Open default dialog');
        this.addButton(new pdg.Rect(535, 447, 735, 487), ids.themedDialog, 'Open themed dialog', themedButton);
    }

    addButton(area, id, text, attributes = null) {
        const button = new Button(this, area, id);
        button.setText(text);
        button.setID(id);
        if (attributes) button.setAttributes(attributes);
        return button;
    }

    addCheckbox(area, text, attributes = null) {
        const checkbox = new Checkbox(this, area);
        checkbox.setString(text);
        if (attributes) checkbox.setAttributes(attributes);
        return checkbox;
    }

    addRadio(area, attributes = null) {
        const radio = new RadioButton(this, area, -1, 3);
        radio.setString(0, 'One');
        radio.setString(1, 'Two');
        radio.setString(2, 'Three');
        if (attributes) radio.setAttributes(attributes);
        return radio;
    }

    getControlAttributes(type) {
        const attributes = new ControlAttributes();
        if (this.useThemedDialog && type === ControlType.Dialog) {
            attributes.stateDrawRoutine(ControlState.Normal, (port, area) => {
                port.drawRect(area, new pdg.Attributes()
                    .fillGradient(area.leftTop(), new pdg.Color(244 / 255, 236 / 255, 1, 1),
                        area.rightBottom(), new pdg.Color(178 / 255, 211 / 255, 1, 1))
                    .lineColor(new pdg.Color(74 / 255, 57 / 255, 145 / 255, 1))
                    .lineThickness(5).roundedCorners(12));
            });
        }
        return attributes;
    }

    buttonClicked(id) {
        if (id === ids.defaultDialog || id === ids.themedDialog) {
            this.useThemedDialog = id === ids.themedDialog;
            new PreviewDialog(this, this.useThemedDialog);
            this.useThemedDialog = false;
        } else if (id !== ids.themedButton) {
            this.status = `Button ${id} clicked`;
        }
    }

    static drawAccentButton(port, area) {
        port.drawRect(area, new pdg.Attributes()
            .fillGradient(area.leftTop(), new pdg.Color(94 / 255, 86 / 255, 220 / 255, 1),
                area.rightBottom(), new pdg.Color(38 / 255, 167 / 255, 190 / 255, 1))
            .lineColor(new pdg.Color(30 / 255, 30 / 255, 80 / 255, 1))
            .lineThickness(2).roundedCorners(10));
    }

    static drawPressedAccentButton(port, area) {
        port.drawRect(area, new pdg.Attributes()
            .fillColor(new pdg.Color(42 / 255, 83 / 255, 135 / 255, 1))
            .lineColor(new pdg.Color(1, 1, 1, 1)).lineThickness(2).roundedCorners(10));
    }
}

class GalleryApplication extends Application {
    setupGraphics() {
        const bounds = pdg.gfx.getScreenBounds();
        const frame = new pdg.Rect(0, 0, 960, 640);
        frame.center(bounds);
        this.mainPort = pdg.gfx.createWindowPort(frame, 'PDG JavaScript Control Gallery', 0);
    }

    preloadResources() {
        this.resourceMgr.openResourceFile('test');
    }

    setupControllers() {
        this.mainController = new GalleryController(null, this, this.mainPort);
    }

    cleanup() {
        if (this.mainController) this.mainController.destroy();
        if (this.mainPort) pdg.gfx.closeGraphicsPort(this.mainPort);
    }
}

new GalleryApplication();

File: test/cxx/app-control-gallery.cpp
// Interactive draw/behavior test for the PDG C++ application controls.

#include "pdg/framework.h"
#include "pdg/app/ControlAttributes.h"
#include "pdg/app/RadioButton.h"

#include <string>

namespace {

constexpr int kWindowWidth = 960;
constexpr int kWindowHeight = 640;

enum ViewId {
    kDefaultButton = 100,
    kDisabledButton,
    kThemedButton,
    kImageButton,
    kDefaultCheckbox,
    kDisabledCheckbox,
    kThemedCheckbox,
    kDefaultRadio,
    kDisabledRadio,
    kThemedRadio,
    kDefaultScrollbar,
    kThemedScrollbar,
    kDefaultDialogButton,
    kThemedDialogButton
};

class GalleryController;

class GalleryCanvas : public pdg::View {
public:
    GalleryCanvas(pdg::Controller* controller, const pdg::Rect& area)
        : View(controller, area) {}

    void drawSelf() override;
};

class DialogLabel : public pdg::View {
public:
    DialogLabel(pdg::Controller* controller, const pdg::Rect& area, std::string text)
        : View(controller, area), mText(std::move(text)) {}

    void drawSelf() override {
        mPort->drawText(mText.c_str(), mViewArea.leftTop() + pdg::Point(12, 28),
            pdg::Attributes().textSize(17).fillColor(PDG_BLACK_COLOR));
    }

private:
    std::string mText;
};

class PreviewDialog : public pdg::Dialog {
public:
    PreviewDialog(pdg::Controller* parent, bool themed)
        : Dialog(parent, 360, 150, dialog_Standard, 1) {
        auto* label = new DialogLabel(this, getDialogRect(), themed
            ? "Custom draw routine for dialog background"
            : "Default PDG dialog background");
        addView(label, 2);

        pdg::Point buttonPoint(getDialogRect().right - 112, getDialogRect().bottom - 44);
        auto* close = new pdg::Button(this,
            pdg::Rect(buttonPoint, 90, 30), 1);
        close->setText("Close");
        addView(close, 1);
    }
};

class GalleryController : public pdg::Controller {
public:
    explicit GalleryController(pdg::Application* app, pdg::Image* exampleImage)
        : Controller(app), mCanvas(nullptr), mUseThemedDialog(false), mClickCount(0),
          mStatus("Click any enabled control") {
        pdg::Rect area = mPort->getDrawingArea();
        mCanvas = new GalleryCanvas(this, area);
        addViewBehind(mCanvas);

        addButton(pdg::Rect(55, 125, 225, 165), kDefaultButton, "Default button", {});
        auto* disabled = addButton(pdg::Rect(55, 180, 225, 220), kDisabledButton,
            "Disabled", {});
        disabled->setEnabled(false);

        pdg::ControlAttributes themedButton;
        themedButton
            .stateDrawRoutine(pdg::ControlState::Normal, drawAccentButton)
            .stateDrawRoutine(pdg::ControlState::Hovered, drawAccentButton)
            .stateDrawRoutine(pdg::ControlState::Pressed, drawPressedAccentButton)
            .stateForeground(pdg::ControlState::Normal, PDG_WHITE_COLOR)
            .stateForeground(pdg::ControlState::Hovered, PDG_WHITE_COLOR)
            .stateForeground(pdg::ControlState::Pressed, PDG_WHITE_COLOR)
            .clickRoutine([this]() { setStatus("Custom button click routine ran"); });
        addButton(pdg::Rect(535, 125, 705, 165), kThemedButton, "Draw routine", themedButton);

        pdg::ControlAttributes imageButton;
        if (exampleImage) imageButton.stateImage(pdg::ControlState::Normal, exampleImage);
        imageButton.stateForeground(pdg::ControlState::Normal, PDG_WHITE_COLOR);
        addButton(pdg::Rect(730, 125, 900, 165), kImageButton, "Image state", imageButton);

        addCheckbox(pdg::Rect(55, 245, 360, 277), kDefaultCheckbox,
            "Default checkbox", {});
        auto* disabledCheckbox = addCheckbox(pdg::Rect(55, 277, 360, 309),
            kDisabledCheckbox, "Disabled checkbox", {});
        disabledCheckbox->setEnabled(false);
        pdg::ControlAttributes themedCheck;
        themedCheck
            .stateForeground(pdg::ControlState::Normal, pdg::Color(45, 68, 140))
            .stateForeground(pdg::ControlState::Selected, pdg::Color(164, 48, 92))
            .clickRoutine([this]() { setStatus("Custom checkbox toggled"); });
        addCheckbox(pdg::Rect(535, 245, 850, 277), kThemedCheckbox,
            "Override text colors", themedCheck);

        addRadio(pdg::Rect(55, 315, 370, 345), kDefaultRadio, {});
        auto* disabledRadio = addRadio(pdg::Rect(130, 355, 430, 385),
            kDisabledRadio, {});
        disabledRadio->setEnabled(false);
        pdg::ControlAttributes themedRadio;
        themedRadio
            .stateForeground(pdg::ControlState::Normal, pdg::Color(36, 94, 72))
            .stateForeground(pdg::ControlState::Selected, pdg::Color(190, 75, 30))
            .clickRoutine([this]() { setStatus("Custom radio selection changed"); });
        addRadio(pdg::Rect(535, 315, 850, 345), kThemedRadio, themedRadio);

		auto* defaultScrollbar = new pdg::Scrollbar(this, pdg::Rect(55, 402, 370, 424),
			pdg::Scrollbar::HORIZONTAL, 35, 10, 110);
		addView(defaultScrollbar, kDefaultScrollbar);
		pdg::ControlAttributes themedScrollbarAttributes;
		themedScrollbarAttributes
			.stateAttributes(pdg::ControlState::Normal,
				pdg::Attributes().fillColor(pdg::Color(224, 216, 246)))
			.stateAttributes(pdg::ControlState::Decrement,
				pdg::Attributes().fillColor(pdg::Color(136, 119, 206)).roundedCorners(4))
			.stateAttributes(pdg::ControlState::Increment,
				pdg::Attributes().fillColor(pdg::Color(136, 119, 206)).roundedCorners(4))
			.stateAttributes(pdg::ControlState::Thumb,
				pdg::Attributes().fillColor(pdg::Color(164, 48, 92)).roundedCorners(6));
		auto* themedScrollbar = new pdg::Scrollbar(this, pdg::Rect(535, 402, 850, 424),
			pdg::Scrollbar::HORIZONTAL, 65, 10, 110);
		themedScrollbar->setAttributes(themedScrollbarAttributes);
		addView(themedScrollbar, kThemedScrollbar);

		addButton(pdg::Rect(55, 447, 255, 487), kDefaultDialogButton,
            "Open default dialog", {});
		addButton(pdg::Rect(535, 447, 735, 487), kThemedDialogButton,
            "Open themed dialog", themedButton);
    }

    pdg::ControlAttributes getControlAttributes(pdg::ControlType type) override {
        pdg::ControlAttributes attributes;
        if (mUseThemedDialog && type == pdg::ControlType::Dialog) {
            attributes.stateDrawRoutine(pdg::ControlState::Normal,
                [](pdg::Port& port, const pdg::Rect& area,
                   const pdg::ControlStateAttributes&) {
                    port.drawRect(area, pdg::Attributes()
                        .fillGradient(area.leftTop(), pdg::Color(244, 236, 255),
                                      area.rightBottom(), pdg::Color(178, 211, 255))
                        .lineColor(pdg::Color(74, 57, 145)).lineThickness(5).roundedCorners(12));
                });
        }
        return attributes;
    }

    bool doLeftClick(const pdg::MouseInfo* mi, pdg::View* view, int id, int part) override {
        bool handled = Controller::doLeftClick(mi, view, id, part);
        if (id == kDefaultDialogButton || id == kThemedDialogButton) {
            mUseThemedDialog = id == kThemedDialogButton;
            new PreviewDialog(this, mUseThemedDialog);
            mUseThemedDialog = false;
            return true;
        }
        if (handled) {
            ++mClickCount;
            if (id == kDefaultButton) setStatus("Default button clicked");
            else if (id == kImageButton) setStatus("Image-backed button clicked");
            else if (id == kDefaultCheckbox) setStatus("Default checkbox toggled");
            else if (id == kDefaultRadio) setStatus("Default radio selection changed");
        }
        return handled;
    }

    const std::string& status() const { return mStatus; }
    int clickCount() const { return mClickCount; }

private:
    static void drawAccentButton(pdg::Port& port, const pdg::Rect& area,
                                 const pdg::ControlStateAttributes&) {
        port.drawRect(area, pdg::Attributes()
            .fillGradient(area.leftTop(), pdg::Color(94, 86, 220),
                          area.rightBottom(), pdg::Color(38, 167, 190))
            .lineColor(pdg::Color(30, 30, 80)).lineThickness(2).roundedCorners(10));
    }

    static void drawPressedAccentButton(pdg::Port& port, const pdg::Rect& area,
                                        const pdg::ControlStateAttributes&) {
        port.drawRect(area, pdg::Attributes().fillColor(pdg::Color(42, 83, 135))
            .lineColor(PDG_WHITE_COLOR).lineThickness(2).roundedCorners(10));
    }

    pdg::Button* addButton(const pdg::Rect& rect, int id, const char* text,
                           const pdg::ControlAttributes& attributes) {
        auto* button = new pdg::Button(this, rect, id);
        button->setText(text);
        button->setAttributes(attributes);
        button->setWantsMouseOvers(true);
        addView(button, id);
        return button;
    }

    pdg::Checkbox* addCheckbox(const pdg::Rect& rect, int id, const char* text,
                               const pdg::ControlAttributes& attributes) {
        auto* checkbox = new pdg::Checkbox(this, rect);
        checkbox->setString(text);
        checkbox->setAttributes(attributes);
        addView(checkbox, id);
        return checkbox;
    }

    pdg::RadioButton* addRadio(const pdg::Rect& rect, int id,
                               const pdg::ControlAttributes& attributes) {
        auto* radio = new pdg::RadioButton(this, rect, -1, 3);
        radio->setString(0, "One");
        radio->setString(1, "Two");
        radio->setString(2, "Three");
        radio->setAttributes(attributes);
        addView(radio, id);
        return radio;
    }

    void setStatus(const char* status) {
        mStatus = status;
        if (mCanvas) mCanvas->draw();
    }

    GalleryCanvas* mCanvas;
    bool mUseThemedDialog;
    int mClickCount;
    std::string mStatus;
};

void GalleryCanvas::drawSelf() {
    auto* gallery = static_cast<GalleryController*>(mController);
    mPort->drawRect(mViewArea, pdg::Attributes().fillColor(pdg::Color(238, 241, 246)));
    mPort->drawText("PDG C++ App Framework Control Gallery", pdg::Point(40, 48),
        pdg::Attributes().textSize(25).textStyle(pdg::textStyle_Bold).fillColor(pdg::Color(30, 38, 55)));
    mPort->drawText("Hover, press, click, toggle, and open both dialogs.", pdg::Point(40, 76),
        pdg::Attributes().textSize(14).fillColor(pdg::Color(70, 78, 92)));

    pdg::Rect defaultPanel(30, 92, 450, 500);
    pdg::Rect overridePanel(510, 92, 930, 500);
    mPort->drawRect(defaultPanel, pdg::Attributes().fillColor(PDG_WHITE_COLOR)
        .lineColor(pdg::Color(185, 191, 202)).roundedCorners(10));
    mPort->drawRect(overridePanel, pdg::Attributes().fillColor(pdg::Color(250, 248, 255))
        .lineColor(pdg::Color(124, 109, 180)).lineThickness(2).roundedCorners(10));
    mPort->drawText("Built-in defaults", pdg::Point(50, 116),
        pdg::Attributes().textSize(17).textStyle(pdg::textStyle_Bold).fillColor(pdg::Color(45, 52, 66)));
    mPort->drawText("Per-control overrides", pdg::Point(530, 116),
        pdg::Attributes().textSize(17).textStyle(pdg::textStyle_Bold).fillColor(pdg::Color(74, 57, 145)));
    mPort->drawText("Disabled:", pdg::Point(55, 375),
        pdg::Attributes().textSize(13).fillColor(pdg::Color(100, 106, 116)));

    std::string status = "Behavior: " + gallery->status() + "   |   handled clicks: "
        + std::to_string(gallery->clickCount());
    mPort->drawRect(pdg::Rect(30, 525, 930, 590), pdg::Attributes()
        .fillColor(pdg::Color(32, 39, 54)).roundedCorners(8));
    mPort->drawText(status.c_str(), pdg::Point(50, 557),
        pdg::Attributes().textSize(16).fillColor(PDG_WHITE_COLOR));
    mPort->drawText("Overrides demonstrated: state colors, image, custom draw routine, click routine.",
        pdg::Point(50, 580), pdg::Attributes().textSize(12).fillColor(pdg::Color(190, 203, 225)));
}

class GalleryApplication : public pdg::Application {
public:
    GalleryApplication() : mController(nullptr), mExampleImage(nullptr) {}

    void initialize(int argc, const char** argv) override {
        (void)argc;
        (void)argv;
        const std::string resourceImage =
            std::string(pdg::OS::getApplicationResourceDirectory()) + "yinyang.png";
        const std::string paths[] = {
            resourceImage,
            "test/data/yinyang.png",
            "../test/data/yinyang.png",
            "../../../../test/data/yinyang.png",
            "yinyang.png"
        };
        for (const std::string& path : paths) {
            try {
                mExampleImage = pdg::Image::createImageFromFile(path.c_str());
                if (mExampleImage) break;
            } catch (...) {}
        }
        mController = new GalleryController(this, mExampleImage);
    }

    void cleanup() override {
        delete mController;
        mController = nullptr;
        if (mExampleImage) {
            mExampleImage->release();
            mExampleImage = nullptr;
        }
    }

private:
    GalleryController* mController;
    pdg::Image* mExampleImage;
};

} // namespace

namespace pdg {

bool Initializer::allowHorizontalOrientation() throw() { return true; }
bool Initializer::allowVerticalOrientation() throw() { return true; }
const char* Initializer::getAppName(bool) throw() { return "PDG Control Gallery"; }
const char* Initializer::getMainResourceFileName() throw() { return nullptr; }

bool Initializer::installGlobalHandlers() throw() {
    new GalleryApplication();
    return true;
}

bool Initializer::getGraphicsEnvironmentDimensions(Rect, Rect, long& width, long& height,
                                                    uint8& depth) throw() {
    width = kWindowWidth;
    height = kWindowHeight;
    depth = 32;
    return false;
}

} // namespace pdg
